frontend/src/components/@common/Button/Button.tsx

```diff
@@ -0,0 +1,167 @@
+// import { HomeIcon } from 'assets/icons';
+import {
+  CSSProperties,
+  ComponentPropsWithRef,
+  ForwardedRef,
+  ReactElement,
+  forwardRef,
+} from 'react';
+
+import { css, styled } from 'styled-components';
+import { RuleSet } from 'styled-components/dist/types';
+
+import type { Color } from 'styles/theme';
+
+// Type
+export const Variant = ['primary', 'text'] as const;
+export type Variant = (typeof Variant)[number];
+
+export const Size = ['small', 'medium', 'large'] as const;
+export type Size = (typeof Size)[number];
+
+export const Align = ['left', 'center', 'right'] as const;
+export type Align = (typeof Align)[number];
+
+export type Props = {
+  variant?: Variant;
+  size?: Size;
+  backgroundColor?: Color;
+  fontColor?: Color;
+  block?: boolean;
+  align?: Align;
+  icon?: ReactElement;
+} & ComponentPropsWithRef<'button'>;
+
+// Component
+const Button = (
+  {
+    children,
+    variant = 'primary',
+    size = 'medium',
+    block = false,
+    align = 'center',
+    icon,
+    ...rest
+  }: Props,
+  ref: ForwardedRef<HTMLButtonElement>,
+) => {
+  return (
+    <S.Button ref={ref} variant={variant} size={size} block={block} align={align} {...rest}>
+      <S.IconTextContainer>
+        {Boolean(icon) && <S.IconWrapper size={size}>{icon}</S.IconWrapper>}
+        <p>{children}</p>
+      </S.IconTextContainer>
+    </S.Button>
+  );
+};
+
+export default forwardRef(Button);
+
+// Styled
+const genVariantStyle = (variant: Required<Props>['variant']): RuleSet<object> => {
+  const styles: Record<typeof variant, ReturnType<typeof genVariantStyle>> = {
+    primary: css`
+      color: ${({ theme }) => theme.color.gray10};
+      background-color: ${({ theme }) => theme.color.gray1};
+      outline: 1px solid ${({ theme }) => theme.color.gray1};
+
+      &:hover {
+        background-color: ${({ theme }) => theme.color.primary};
+        color: ${({ theme }) => theme.color.gray2};
+      }
+
+      &:focus {
+        box-shadow: 0 0 0 3px ${({ theme }) => theme.color.primary};
+      }
```

Comment:
```suggestion
    primary: css`
      ${({ theme }) => `
        color: ${theme.color.gray10};
  
        background-color: ${theme.color.gray1};
        outline: 1px solid ${theme.color.gray1};
  
        &:hover {
          background-color: ${theme.color.primary};
          color: ${theme.color.gray2};
        }
  
        &:focus {
          box-shadow: 0 0 0 3px ${theme.color.primary};
        }
    `}
```
이거 theme 매번 가져오는 것보다 이렇게 쓰면 어떨까요?

Reply:
아 저런방법이 있군요?!?!
안그래도 쓰면서 되게 복잡하다고 생각했는데.. 
바로 반영하겠습니다😆